spec/haskell/src/SEL4/Object/SchedContext.lhs

```diff
@@ -345,15 +345,46 @@ This module uses the C preprocessor to select a target architecture.
 > refillHeadOverlappingLoop scPtr =
 >        whileLoop (const (fromJust . runReaderT (refillHeadOverlapping scPtr))) (const (mergeRefills scPtr)) ()
 
+> headTimeBuffer :: Ticks -> KernelR Bool
+> headTimeBuffer usage = do
+>     scPtr <- readCurSc
+>     sc <- readSchedContext scPtr
+>     return $ rAmount (refillHd sc) <= usage && (maxBound :: Word64) - rTime (refillHd sc) >= 5 * usToTicks maxPeriodUs
+
+> handleOverrunLoopBody :: Ticks -> Kernel Ticks
+> handleOverrunLoopBody usage = do
+>     scPtr <- getCurSc
+>     single <- refillSingle scPtr
+>     sc <- getSchedContext scPtr
+
+>     usage' <- return (usage - rAmount (refillHd sc))
+
+>     if single
+>        then updateRefillHd scPtr $ \r -> r { rTime = rTime r + scPeriod sc}
+>        else do
+>          old_head <- refillPopHead scPtr
+>          let new = old_head { rTime = rTime old_head + scPeriod sc}
+>          scheduleUsed scPtr new
+
+>     return usage'
+
+> handleOverrunLoop :: Ticks -> Kernel Ticks
+> handleOverrunLoop usage =
+>     whileLoop (\usage -> fromJust . runReaderT (headTimeBuffer usage)) (\usage -> handleOverrunLoopBody usage) usage
+
 > refillBudgetCheck :: Ticks -> Kernel ()
 > refillBudgetCheck usage = do
 >     scPtr <- getCurSc
+
 >     sc <- getSchedContext scPtr
 >     assert (scRefillMax sc > 0) "CurSc should be active"
```

Comment:
Unrelated to this PR and probably not necessary, but it might be slightly more consistent to use `scActive` here (and maybe readerfy it).
```suggestion
>     active <- scActive scPtr
>     assert active "CurSc should be active"
```

Reply:
I decided against the readificationizing of `scActive`. There were a few too many lemmas that it would have touched, and it might be best to handle this in some future update that addresses reading/getting more generally